File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/BaseViewModelFragment.java
package mx.dev.franco.automusictagfixer.ui;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.AndroidSupportInjection;
import dagger.android.support.HasSupportFragmentInjector;

/**
 * Base fragment that abstract the common functionality for fragments
 * that inherits from it.
 *
 * @author Franco Castillo
 */
public abstract class BaseViewModelFragment<ViewModel> extends BaseFragment implements HasSupportFragmentInjector {
    @Inject
    protected DispatchingAndroidInjector<Fragment> fragmentDispatchingAndroidInjector;
    @Inject
    protected AndroidViewModelFactory androidViewModelFactory;
    protected ViewModel mViewModel;

    /**
     * Called when a fragment is first attached to its context.
     * {@link #onCreate(Bundle)} will be called after this.
     */
    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = getViewModel();
    }

    protected abstract ViewModel getViewModel();

    protected void loading(boolean isLoading){ }

    @Override
    public AndroidInjector<Fragment> supportFragmentInjector() {
        return fragmentDispatchingAndroidInjector;
    }
}

File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/BaseViewModelFragment.kt
package mx.dev.franco.automusictagfixer.ui

import android.content.Context
import android.os.Bundle
import androidx.lifecycle.ViewModel
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import dagger.android.support.AndroidSupportInjection
import javax.inject.Inject

/**
 * Base fragment that abstract the common functionality for fragments
 * that inherits from it.
 *
 * @author Franco Castillo
 */
abstract class BaseViewModelFragment<VM: ViewModel> : BaseFragment(), HasAndroidInjector {

    @Inject
    lateinit var fragmentDispatchingAndroidInjector: DispatchingAndroidInjector<Any>

    @Inject
    lateinit var androidViewModelFactory: AndroidViewModelFactory

    protected lateinit var viewModel: VM

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = obtainViewModel()
    }

    abstract fun obtainViewModel(): VM

    open fun loading(isLoading: Boolean) {}

    override fun androidInjector(): AndroidInjector<Any> = fragmentDispatchingAndroidInjector
}